### PR TITLE
fix: replace use of deprecated methods

### LINKED ISSUE #899: Prepare the rules for ESLint v9
### Plugin version

n/a

### What problem do you want to solve?

- [Preparing your custom rules for ESLint v9.0.0](https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/)

### Your take on the correct solution?

Apply the changes mentioned in the linked migration docs so the custom rules are adapted to ESLint v9 and the

### Anything else?

_No response_

### Do you want to submit a pull request to implement this change?

Yes

### LINKED ISSUE #924: eslint v9 - context.getScope is not a function - API change
### Have you read the Troubleshooting section?

Yes

### Plugin version

v6.3.0

### ESLint version

v9..9.0

### Node.js version

20.16.0

### package manager and version

yarn v1

### Operating system

macOS

### Bug description

Example error:

```
TypeError: context.getScope is not a function
Occurred while linting ./examples/example-web-frontend/src/hello-world.test.tsx:39
Rule: "testing-library/await-async-utils"
```

### Steps to reproduce

run eslint in terminal or ide using eslint v9.9.0 and latest plugin.

### Error output/screenshots

```shell
Rule: "testing-library/await-async-utils"
    at getInnermostFunctionScope (/project/node_modules/eslint-plugin-testing-library/dist/node-utils/index.js:205:71)
    at getInnermostReturningFunction (/project/node_modules/eslint-plugin-testing-library/dist/node-utils/index.js:331:27)
    at detectAsyncUtilWrapper (/project/node_modules/eslint-plugin-testing-library/dist/rules/await-async-utils.js:32:82)
    at CallExpression Identifier (/project/node_modules/eslint-plugin-testing-library/dist/rules/await-async-utils.js:86:21)
    at enhancedRuleInstructions.<computed> (/project/node_modules/eslint-plugin-testing-library/dist/create-testing-library-rule/detect-testing-library-utils.js:528:105)
    at ruleErrorHandler (/project/node_modules/eslint/lib/linter/linter.js:1124:48)
    at /Users/dgreene/medallia/magic-fun/node_modules/eslint/lib/linter/safe-emitter.js:45:58
```

### ESLint configuration

Not the whole config, just the part regarding test files.
The project is using `typescript-eslint@v8.0.1`.

```ts
import globals from 'globals';
import testingLibrary from 'eslint-plugin-testing-library';
import { config } from 'typescript-eslint';
import jest from 'eslint-plugin-jest';
import jestDom from 'eslint-plugin-jest-dom';

export const testConfig = config({
  name: 'eslint-config-test-files',

  files: [
    '**/*.test.?(disabled.)?(m)[jt]s?(x)',
    '**/*.spec.?(m)[jt]s?(x)',
    '**/demo/**/*.[jt]s?(x)',
    '**/demo-sandbox/**/*.[jt]s?(x)',
    '**/__mocks__/**/*.[jt]s?(x)',
    '**/test/**/*.[jt]s?(x)',
    '**/fixtures/**/*.[jt]s?(x)'
  ],

  languageOptions: {
    globals: { ...globals.jest, jsdom: true }
  },

  plugins: {
    jest,
    'jest-dom': jestDom,
    'testing-library': testingLibrary
  },

  rules: {
    ...jest.configs['flat/recommended'].rules,
    ...jestDom.configs['flat/recommended'].rules,
    ...testingLibrary.configs['flat/react'].rules
  }
});
```

### Rule(s) affected

Seems like all of them.

### Anything else?

It seems that `getScope` was moved from `context` to `sourceCode`

https://eslint.org/docs/latest/use/troubleshooting/v9-rule-api-changes

https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/#context.getscope()

### Do you want to submit a pull request to fix this bug?

No

### PR DESCRIPTION
**Pull request by G-Rath**
## Checks

- [x] I have read the [contributing guidelines](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/CONTRIBUTING.md).

## Changes

- introduced compat utilities that enable support for ESLint v9
- replaced usages of `context` functions that were removed in v9 in favor of compat utils

## Context

Resolves #924
Resolves #899

### CODE REVIEW

**Comment by G-Rath** [comment]
#926 proves that this works with ESLint v7, v8, and v9 so I think this should be safe to merge as-is and then we'll be one step closer to supporting ESLint v9 officially

**Comment by G-Rath** [comment]
@Belco90 @MichaelDeBoey would it be possible to get this and my follow-up PRs progressed, given that they can be released in a minor version?

**Comment by Belco90** [comment]
Thanks for your contribution, and sorry for the inactivity. I'll try to get this merged during the week.

**Comment by G-Rath** [comment]
@Belco90 no worries, let me know if there's anything I can do to help - just in case you've not seen it, I've got a few pull requests open which include ones that ultimately lead to allowing ESLint v9 which are chained off of each other.

In order they are:
  - #925 (this one)
  - #927 (optional)
  - #926
  - #928

I should be able to have them rebased within minutes of you landing the prior ones unless I'm sleep, but I can also do a single big PR if you're prefer.

Let me know what works best for you, and also if you have questions or want something done differently

**Comment by Belco90** [comment]
@G-Rath Amazing, thank you so much! I'm going to merge some other chore PRs first (like your editorconfig PR), then we can rebase this one and start reviewing in the order you mentioned.

File: lib/utils/compat.ts
Comment (Belco90) [inline lib/utils/compat.ts:1]: Great idea! I bet this module can be removed once we release the next major only compatible with ESLint v9?

**Review by Belco90 — APPROVED** [review]
LGTM! I'll wait until it's rebased to merge it (hopefully the failing test gets fixed). Thanks again!

File: lib/utils/compat.ts
Comment (G-Rath) [reply lib/utils/compat.ts:1]: yup exactly (though hopefully that particular major won't be for a while longer since the ecosystem is still very much catching up to v9 😅)

**Comment by G-Rath** [comment]
@Belco90 this is good to go now!

**Review by MichaelDeBoey — APPROVED** [review]

**Comment by Belco90** [comment]
@G-Rath Ah sorry, this must be reformatted after I merged the prettier v3 upgrade.

**Comment by G-Rath** [comment]
@Belco90 done 🙂

**Review by Belco90 — APPROVED** [review]

**Pull request merged**
